### DISCUSSION TITLE
Setting Request Headers

### DISCUSSION THREAD

**Discussion by albertoramires** [keathley/wallaby#705] [Q&A]
Is there any way to set headers in Chrome? The same headers could apply to all requests.

**Comment by mhanberg** [comment]
To clarify, do you mean on the requests to the web driver or like, when you click links in your app?

**Reply by albertoramires** [reply]
Neither, but it somewhat falls under "click links in your app", all the requests are API endpoint requests so it's just one JSON visit, the headers would need to be on this request.
Unfortunately I need to use Chrome, I don't see any other way to bypass Cloudflare protections.

**Reply by mhanberg** [reply]
So you want to set a header on the GET request that the browser would make? or are you making JS requests from within Chrome? or are these from a <form> POST?

**Reply by albertoramires** [reply]
GET request the browser would make, no JS involved.

**Reply by mhanberg** [reply]
This might be useful: https://stackoverflow.com/questions/30813287/add-httpheader-to-selenium-chrome-webdriver-in-c-sharp#37662415

Basically you load a Chrome Extension that allows for setting custom headers and then run a script to configure it.

Other than that, I'm not sure there is a way to do it natively.

**Reply by mhanberg** [reply]
But if possible, you can still do a GET fetch request in JS and set headers there.

**Reply by albertoramires** [reply]
>Basically you load a Chrome Extension that allows for setting custom headers and then run a script to configure it.

Is it possible to inject extensions into chrome through Wallaby though?

**Reply by mhanberg** [reply]
https://chromedriver.chromium.org/capabilities

You should be able to pass them into the extensions key of the chromeOptions object that is passed in the capabilities object.

**Reply by albertoramires** [reply]
Thank you so much, I'll give it a shot.

**Reply by albertoramires** [reply]
I really can't seem to load the extension through `extensions`, tried both full path and relative.

**Reply by albertoramires** [reply]
Nevermind, just noticed that it needs to load the file into base64. Appears to be working! ✨
If anyone has the same issue: `extensions: [Path.absname("modheader.crx") |> File.read!() |> Base.encode64()]`

**Reply by mhanberg** [reply]
Great to hear!
